consensus/src/main/kotlin/maru/consensus/validation/BlockValidator.kt

```diff
@@ -0,0 +1,123 @@
+import maru.core.BeaconBlockHeader
+import maru.core.HashUtil
+import maru.database.Database
+import maru.executionlayer.client.ExecutionLayerClient
+import maru.executionlayer.extensions.hasValidExecutionPayload
+import maru.serialization.rlp.bodyRoot
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+
+interface BlockValidator {
+  fun validateBlock(
+    block: BeaconBlock,
+    prevBlockHeader: BeaconBlockHeader,
+  ): SafeFuture<Result<Boolean>>
+
+  fun validateBlock(block: BeaconBlock): SafeFuture<Result<Boolean>>
+}
+
+class BlockValidationException : Exception {
+  constructor(message: String) : super(message)
+  constructor(message: String, cause: Throwable) : super(message, cause)
+}
+
+class BlockValidatorImpl(
+  private val validatorProvider: ValidatorProvider,
+  private val sealVerifier: SealVerifier,
+  private val executionLayerClient: ExecutionLayerClient,
+  private val database: Database,
+) : BlockValidator {
+  companion object {
+    fun invalid(message: String): Result<Boolean> = Result.failure(BlockValidationException("Invalid block: $message"))
+
+    fun invalid(
+      message: String,
+      cause: Throwable,
+    ): Result<Boolean> =
+      Result.failure(BlockValidationException("Invalid block: $message" + ", cause: ${cause.message}", cause))
+
+    fun valid(): Result<Boolean> = Result.success(true)
+  }
+
+  override fun validateBlock(
+    block: BeaconBlock,
+    prevBlockHeader: BeaconBlockHeader,
+  ): SafeFuture<Result<Boolean>> {
+    if (block.beaconBlockHeader.number != prevBlockHeader.number + 1u) {
+      return SafeFuture.completedFuture(
+        invalid("Block number is not the next block number"),
+      )
+    }
+
+    if (block.beaconBlockHeader.timestamp <= prevBlockHeader.timestamp) {
+      return SafeFuture.completedFuture(
+        invalid("Block timestamp is not greater than previous block timestamp"),
+      )
+    }
+
+    if (validatorProvider.getProposerForBlock(block.beaconBlockHeader) != block.beaconBlockHeader.proposer) {
+      return SafeFuture.completedFuture(invalid("Proposer is not expected proposer"))
+    }
+
+    if (!block.beaconBlockHeader.parentRoot.contentEquals(prevBlockHeader.hash)) {
+      return SafeFuture.completedFuture(invalid("Parent root does not match"))
+    }
+
+    val beaconBodyRoot = HashUtil.bodyRoot(block.beaconBlockBody)
+
+    if (!block.beaconBlockHeader.bodyRoot.contentEquals(beaconBodyRoot)) {
+      return SafeFuture.completedFuture(invalid("Body root does not match"))
+    }
+
+    val prevBlockValidators = validatorProvider.getValidatorsForBlock(prevBlockHeader)
+    if (block.beaconBlockBody.prevCommitSeals.size < Math.ceilDiv(prevBlockValidators.size, 2)) {
+      return SafeFuture.completedFuture(
+        invalid("Not enough commit seals for previous block"),
+      )
+    }
+
+    for (seal in block.beaconBlockBody.prevCommitSeals) {
+      val sealVerificationResult = sealVerifier.verifySeal(seal, prevBlockHeader)
```

Comment:
Similar to @Filter94 comment above, using the `prevBlockHeader` won't give the correct set of validators to verify the seal against. We always include validators needed to build the next block. So the `prevCommitSeals` used the validator set from `prevBlockHeader.parentRoot`